Implement connection string

shaan1337 opened the pull request:
Fixes #41

Note: This PR has been opened against `use-https-option` branch so that it's retargeted to `master` after the former is merged.

Examples:
```
single node:
EventStoreClientSettings.Create("esdb://localhost"); //with hostname only
EventStoreClientSettings.Create("esdb://localhost:2113"); //with hostname and port
EventStoreClientSettings.Create("esdb://user:pass@localhost:2113"); //with username and password
EventStoreClientSettings.Create("esdb://user:pass@localhost:2113/"); //extra forward slash also works
EventStoreClientSettings.Create("esdb://user:pass@localhost:2113?tlsVerifyCert=false"); //disables server certificate validation
EventStoreClientSettings.Create("esdb://user:pass@localhost:2113?tls=false"); //connects over http instead of https

cluster:
EventStoreClientSettings.Create("esdb://host1,host2,host3"); //with hostnames only
EventStoreClientSettings.Create("esdb://host1:1234,host2:4321,host3:3231"); //with hostnames and port
EventStoreClientSettings.Create("esdb://user:pass@host1:1234,host2:4321,host3:3231?nodePreference=follower"); //with username and password
EventStoreClientSettings.Create("esdb://host1,host2,host3?tls=false"); //connects over http instead of https
EventStoreClientSettings.Create("esdb://host1,host2,host3?tlsVerifyCert=false"); //disables server certificate validation
```

All parameters:
```
ConnectionName / string
MaxDiscoverAttempts / int
DiscoveryInterval / int (milliseconds)
GossipTimeout / int (milliseconds)
NodePreference / string
Tls / bool
TlsVerifyCert / bool
OperationTimeout / int (milliseconds)
ThrowOnAppendFailure / bool
```

thefringeninja commented on src/EventStore.Client/ConnectionStringParseException.cs:
```diff
@@ -0,0 +1,37 @@
+using System;
+
+namespace EventStore.Client {
+	public class ConnectionStringParseException : Exception {
```
Can we put each of these into their own file? It would make finding it on github easier.

shaan1337 replied:
yup, it has been fixed in this commit: da3a713c7f99e7a1297bfcd9452b37e7a2933be6

thanks

pgermishuys commented:
Let's try and marry up the units to that of the server.

shaan1337 commented:
@hayley-jean good catch, thanks for the feedback, I've addressed these issues in these 2 commits:
f4ffa9c411dac609c12267c8607302cfbab615c6
bfad20116e8548d083a0c09dee787b90f7ca78fb

hayley-jean commented on src/EventStore.Client/EventStoreClientSettings.ConnectionString.cs:
```diff
@@ -0,0 +1,246 @@
+				var questionMarkIndex = connectionString.IndexOf(QuestionMark, currentIndex, StringComparison.Ordinal);
+				if ((questionMarkIndex == -1 && slashIndex != connectionString.Length - Slash.Length)
+				    || (questionMarkIndex != -1 && questionMarkIndex != currentIndex))
+					throw new ConnectionStringParseException($"The connection string may contain only a question mark (?) after specifying the hosts and a forward slash (/) but the following character was found: '{connectionString[currentIndex]}'");
```
I think this error might be a bit confusing, is there a way we could reword it?
The best I can come up with right now is:

```suggestion
					throw new ConnectionStringParseException($"The connection string must contain a question mark (?) following the hosts if more arguments are specified. The following character was found instead: '{connectionString[currentIndex]}'");
```

shaan1337 replied:
@hayley-jean  this looks better, I've committed the changes, thanks

pgermishuys commented:
Hey @shaan1337, good work on this!
Is there any particular reason why we are requiring a `/` and `?` after the hosts to indicate extra parameters? Wouldn't a single character such as a `;` give it a better user experience?

shaan1337 commented:
@pgermishuys yes, the connection string mimics a URL as much as possible and ? is part of the syntax: https://en.wikipedia.org/wiki/URL#Syntax

Right now we enforce a ? just after the / but the path could later on point to a specific resource, e.g if we later on support multi-tenancy we could specify the database name in the path:
EventStoreClientSettings.Create("esdb://localhost/mydatabase1?tls=false");

shaan1337 commented:
@pgermishuys @hayley-jean @thefringeninja 
I've added a new commit (52ffeab5a98fe4b1f7d103ddc0f12850a0353a87) to make the path optional and removed the last 2 commits. It's up for review again (except some unrelated flaky test is failing intermittently)
